osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs

```diff
@@ -0,0 +1,232 @@
+
+        private readonly Dictionary<LifetimeEntry, DrawableCaughtObject> drawableMap = new Dictionary<LifetimeEntry, DrawableCaughtObject>();
+
+        private readonly HashSet<CaughtObjectEntry> aliveStackedObjects = new HashSet<CaughtObjectEntry>();
+        private readonly Dictionary<CaughtObjectEntry, CaughtObjectEntry> dropEntryMap = new Dictionary<CaughtObjectEntry, CaughtObjectEntry>();
+
+        private readonly DrawablePool<DrawableCaughtFruit> caughtFruitPool;
+        private readonly DrawablePool<DrawableCaughtBanana> caughtBananaPool;
+        private readonly DrawablePool<DrawableCaughtDroplet> caughtDropletPool;
+
+        /// <summary>
+        /// The randomness used to compute position in stack.
+        /// It is incremented for each <see cref="Add"/> call and decremented for each <see cref="Remove"/> call to make a replay consistent.
+        /// </summary>
+        private int randomSeed = 1;
+
+        public CaughtObjectContainer(Container<DrawableCaughtObject> droppedObjectTarget)
+        {
+            this.droppedObjectTarget = droppedObjectTarget;
+
+            InternalChildren = new Drawable[]
+            {
+                caughtFruitPool = new DrawablePool<DrawableCaughtFruit>(50),
+                caughtBananaPool = new DrawablePool<DrawableCaughtBanana>(100),
+                // less capacity is needed compared to fruit because droplet is not stacked
+                caughtDropletPool = new DrawablePool<DrawableCaughtDroplet>(25),
+                StackedObjectContainer = new Container<DrawableCaughtObject>()
+            };
+
+            lifetimeManager.EntryBecameAlive += entryBecameAlive;
+            lifetimeManager.EntryBecameDead += entryBecameDead;
+        }
+
+        public void Add(CaughtObjectEntry entry)
+        {
+            randomSeed++;
+
+            lifetimeManager.AddEntry(entry);
+
+            if (entry.State != CaughtObjectState.Stacked) return;
+
+            // `DropStackedObjects` may be called before lifetime update.
+            if (entry.LifetimeStart <= Time.Current)
+                aliveStackedObjects.Add(entry);
+
+            var dropEntry = new CaughtObjectEntry(CaughtObjectState.Dropped, entry.PositionInStack, entry)
+            {
+                LifetimeStart = double.PositiveInfinity,
+                LifetimeEnd = double.PositiveInfinity
+            };
+
+            lifetimeManager.AddEntry(dropEntry);
+            dropEntryMap[entry] = dropEntry;
+        }
+
+        public void Remove(CaughtObjectEntry entry)
+        {
+            randomSeed--;
+
+            lifetimeManager.RemoveEntry(entry);
+            removeDrawable(entry);
+
+            aliveStackedObjects.Remove(entry);
+
+            if (!dropEntryMap.TryGetValue(entry, out var dropEntry)) return;
+
+            dropEntryMap.Remove(entry);
+
+            lifetimeManager.RemoveEntry(dropEntry);
+            removeDrawable(dropEntry);
```

Comment:
All of these fields are really hard to keep track of (what was added where and what needs to be cleaned up from where).

Would probably factor out four symmetrical methods, something like:

```csharp
public void addImmediateEntry(CaughtObjectEntry entry)
{
	lifetimeManager.AddEntry(entry);
	if (entry.State == CaughtObjectState.Stacked)
		addDelayedDropEntry(entry);
}

public void removeImmediateEntry(CaughtObjectEntry entry)
{
	lifetimeManager.RemoveEntry(entry);
	if (entry.State == CaughtObjectState.Stacked)
		removeDelayedDropEntry(entry);
}

public void addDelayedDropEntry(CaughtObjectEntry entry)
{
	var dropEntry = new CaughtObjectEntry(CaughtObjectState.Dropped, entry.PositionInStack, entry)
	{
		LifetimeStart = double.PositiveInfinity,
		LifetimeEnd = double.PositiveInfinity
	};

	lifetimeManager.AddEntry(dropEntry);
	dropEntryMap[entry] = dropEntry;
}

public void removeDelayedDropEntry(CaughtObjectEntry entry)
{
	dropEntryMap.Remove(entry, out var dropEntry);
	lifetimeManager.RemoveEntry(dropEntry);
}
```

I would hope that something can be done to eliminate the `aliveStackedObjects` adds/deletes here, such that that logic can be entirely located in `entryBecame{Alive,Dead}`.

Also not sure why `Remove()` is cleaning up drop entry drawables but `Add()` never adds them. If that is indeed required for some reason, then I'd leave a comment in `addDelayedDropEntry` as to why.